bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -1457,30 +1464,34 @@ void PreProcessor::registerAndHandleClientPreProcessReqOnNonPrimary(const string
   }
 }
 
+uint32_t PreProcessor::getBufferOffset(uint16_t clientId, ReqId reqSeqNum, uint16_t reqOffsetInBatch) const {
+  return (clientId - numOfReplicas_ - numOfInternalClients_) * clientMaxBatchSize_ + reqOffsetInBatch;
+}
+
 const char *PreProcessor::getPreProcessResultBuffer(uint16_t clientId, ReqId reqSeqNum, uint16_t reqOffsetInBatch) {
-  // Allocate on first use buffers scheme:
+  // Buffers structure scheme:
   // |first client's first buffer|...|first client's last buffer|......
   // |last client's first buffer|...|last client's last buffer|
   // First client id starts after the last replica id.
   // First buffer offset = numOfReplicas_ * batchSize_
   // The number of buffers per client comes from the configuration parameter clientBatchingMaxMsgsNbr.
-  const auto bufferOffset =
-      (clientId - numOfReplicas_ - numOfInternalClients_) * clientMaxBatchSize_ + reqOffsetInBatch;
-  LOG_TRACE(logger(), KVLOG(clientId, reqSeqNum, reqOffsetInBatch, bufferOffset, reqSeqNum));
-  char *buf = nullptr;
-  if (!preProcessResultBuffers_[bufferOffset].first) {
-    // sizeof(uint64_t) -> block id of conflict detection optimization
-    buf = new char[maxPreExecResultSize_ + sizeof(uint64_t)];
-    {
-      std::unique_lock lock(resultBufferLock_);
-      if (!preProcessResultBuffers_[bufferOffset].first) {
-        preProcessResultBuffers_[bufferOffset].second = Sliver(buf, maxPreExecResultSize_);
-        preProcessResultBuffers_[bufferOffset].first = true;
-      } else
-        delete[] buf;
-    }
+  const auto bufferOffset = getBufferOffset(clientId, reqSeqNum, reqOffsetInBatch);
+  std::unique_lock lock(preProcessResultBuffers_[bufferOffset]->mutex);
```

Comment:
If I understand, the mutex protects the buffer, but on the other hand, the method returns a pointer to the buffer i.e. the buffer is not protected and can be accessed concurrently, no?

Reply:
We need to protect only the buffer allocation. I cannot think of a scenario when we modify the same buffer from different threads.